### PR TITLE
Companion for EPM duplicate submissions

### PR DESCRIPTION
**Pull request by kianenigma**
https://github.com/paritytech/substrate/pull/12237

### CODE REVIEW

**Review by ordian — APPROVED** [review]

File: utils/staking-miner/src/opts.rs [hunk 1]
```diff
@@ -196,6 +199,9 @@ impl FromStr for SubmissionStrategy {
 			Self::IfLeading
 		} else if s == "always" {
 			Self::Always
+		} else if s.starts_with("no-worse-than ") {
```
Comment (niklasad1) [inline utils/staking-miner/src/opts.rs:202]:
```suggestion
		} else if let Some(percent) = s.strip_prefix("no-worse-than ") {
```

File: utils/staking-miner/src/monitor.rs [hunk 2]
```diff
@@ -104,34 +121,16 @@ async fn ensure_no_better_solution<T: EPM::Config, B: BlockT>(
...
+	// we check the queue here as well. Could be checked elsewhere.
+	if indices.len() as u32 >= max_submissions {
+		return Err(Error::<T>::QueueFull)
 	}
 
-	if scores == max_submissions {
-		log::warn!(target: LOG_TARGET, "The submissions queue is full");
-	}
+	// default score is all zeros, any score is better than it.
+	let best_score = indices.last().map(|(score, _, _)| *score).unwrap_or_default();
```
Comment (niklasad1) [inline utils/staking-miner/src/monitor.rs:130]: is this `Vec` sorted by score?

**Review by niklasad1 — APPROVED** [review]

File: utils/staking-miner/src/monitor.rs (hunk 2 above)
Comment (kianenigma) [reply utils/staking-miner/src/monitor.rs:130]: indeed it is.

File: utils/staking-miner/src/monitor.rs [hunk 3]
```diff
@@ -325,10 +329,10 @@ macro_rules! monitor_cmd_for { ($runtime:tt) => { paste::paste! {
 
 			// Run the calls in parallel and return once all has completed or any failed.
 			if let Err(err) = tokio::try_join!(
-				flatten(ensure_no_better_fut),
+				flatten(ensure_strategy_met_fut),
 				flatten(ensure_signed_phase_fut),
 			) {
-				log::debug!(target: LOG_TARGET, "Skipping to submit at block {}; {}", at.number, err);
+				log::error!(target: LOG_TARGET, "Skipping to submit at block {}; {}", at.number, err);
```
Comment (kianenigma) [inline utils/staking-miner/src/monitor.rs:335]:
```suggestion
				log::debug!(target: LOG_TARGET, "Skipping to submit at block {}; {}", at.number, err);
```

File: utils/staking-miner/src/main.rs [hunk 4]
```diff
@@ -420,17 +421,37 @@ fn mine_dpos<T: EPM::Config>(ext: &mut Ext) -> Result<(), Error<T>> {
...
+		);
+	};
+
+	if print {
+		do_print();
+	}
 
-	if linked_version != on_chain_version {
+	// we relax the checking here a bit, which should not cause any issues in production (a chain
+	// that messes up its spec name is highly unlikely), but it allows us to do easier testing.
```
Comment (ggwpez) [inline utils/staking-miner/src/main.rs:450]: About messed up spec name; I accidentally ran the try-runtime with the wrong `--chain` argument.  
So I would run Kusama migrations on a Polkadot node and it worked :see_no_evil:   
Could we maybe add an extra flag (not in this MR) that controls whether a different spec name should panic?

**Review by ggwpez — APPROVED** [review]

File: utils/staking-miner/src/main.rs (hunk 4 above)
Comment (kianenigma) [reply utils/staking-miner/src/main.rs:450]: we already have something very similar: https://paritytech.github.io/substrate/master/try_runtime_cli/struct.SharedParams.html#

**Review by bkchr — APPROVED** [review]

**Pull request merged**